piecrust/src/store/memory.rs

```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
+    pub fn new() -> io::Result<Self> {
+        let len = MIN_MEM_SIZE;
+
+        let ptr = mmap(
+            ptr::null_mut(),
+            MAX_MEM_SIZE,
+            PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE,
+            -1,
+            0,
+        )?;
+
+        // An offset guard page is added to optimize reads and writes.
+        Ok(Self { ptr, len })
+    }
+
+    /// Maps the given file, creating it if it doesn't exist.
+    ///
+    /// Memories mapped from a file are mapped in two steps. In the first step
+    /// the file is mmapped copy-on-write, reserving a region as large as
+    /// `MAX_MEM_SIZE`. In the second step an overlapping region is mapped to
+    /// physical memory, at an offset of file length, using `MAP_FIXED`. This
+    /// ensures the memory is free to grow beyond the boundaries of the file
+    /// into physical memory, in one contiguous memory region.
+    ///
+    /// # Errors
+    /// If the file already exists, and it has a length that is not a multiple
+    /// of WASM_PAGE_SIZE, the function will error.
+    pub fn map<P: AsRef<Path>>(file: P) -> io::Result<Self> {
+        let file = fs::OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .open(file)?;
+
+        let file_len = file.metadata()?.len();
+        if file_len % WASM_PAGE_SIZE as u64 != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "File length is not a multiple of WASM page size",
+            ));
+        }
+
+        // Set the length of the file to the minimum if too small.
+        let min_len = MIN_MEM_SIZE as u64;
+        if file_len < min_len {
+            file.set_len(min_len)?;
+        }
+
+        let fd = file.as_raw_fd();
+        let file_len = cmp::max(file_len, min_len) as usize;
+
+        let ptr = mmap(
+            ptr::null_mut(),
+            MAX_MEM_SIZE,
+            PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_NORESERVE,
+            fd,
+            0,
+        )?;
+
+        let growth_ptr = unsafe { ptr.add(file_len) };
+
+        mmap(
+            growth_ptr,
+            MAX_MEM_SIZE - file_len,
+            PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE | MAP_FIXED,
+            -1,
+            0,
+        )?;
+
+        Ok(Self { ptr, len: file_len })
+    }
+
+    /// Grow the map by the specified number of bytes.
+    ///
+    /// Growing in this scheme effectively means incrementing the length.
+    pub fn grow_by(&mut self, delta_bytes: usize) -> io::Result<()> {
+        let new_len = self.len + delta_bytes;
+
+        if new_len > MAX_MEM_SIZE {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Tried growing past the maximum number of bytes: {MAX_MEM_SIZE}"),
+            ));
+        }
+
+        self.len = new_len;
```

Comment:
Does it worth to check if the len is still a multiple of wasm page?

Reply:
We could do this as a debug assertion, but I wouldn't waste runtime on release mode.